lib/northbound_grpc.cpp

```diff
@@ -37,1315 +39,1276 @@
 
 #define GRPC_DEFAULT_PORT 50051
 
-static void *grpc_pthread_start(void *arg);
-
 /*
  * NOTE: we can't use the FRR debugging infrastructure here since it uses
  * atomics and C++ has a different atomics API. Enable gRPC debugging
  * unconditionally until we figure out a way to solve this problem.
  */
-static bool nb_dbg_client_grpc = 1;
+static bool nb_dbg_client_grpc = 0;
+
+static struct thread_master *main_master;
 
 static struct frr_pthread *fpt;
 
-/* Default frr_pthread attributes */
-static const struct frr_pthread_attr attr = {
-	.start = grpc_pthread_start,
-	.stop = NULL,
+#define grpc_debug(...)                                                        \
+	do {                                                                   \
+		if (nb_dbg_client_grpc)                                        \
+			zlog_debug(__VA_ARGS__);                               \
+	} while (0)
+
+// ------------------------------------------------------
+//                      New Types
+// ------------------------------------------------------
+
+enum CallState { CREATE, PROCESS, MORE, FINISH, DELETED };
+const char *call_states[] = {"CREATE", "PROCESS", "MORE", "FINISH", "DELETED"};
+
+struct candidate {
+	uint64_t id;
+	struct nb_config *config;
+	struct nb_transaction *transaction;
 };
 
-enum CallStatus { CREATE, PROCESS, FINISH };
+class Candidates
+{
+      public:
+	~Candidates(void)
+	{
+		// Delete candidates.
+		for (auto it = _cdb.begin(); it != _cdb.end(); it++)
+			delete_candidate(&it->second);
+	}
+
+	struct candidate *create_candidate(void)
+	{
+		uint64_t id = ++_next_id;
+		assert(id); // TODO: implement an algorithm for unique reusable
```

Comment:
Is there a problem if this number wraps? What is the worst case?

I'm unfamiliar with this code, but `assert` is too drastic, couldn't we just check if the ID is already in use?

Reply:
The old code was the same but a uint32_t I didn't try and fix it, but I made it a lot more hard to hit. :)

(((((18446744073709551615/1000)/60)/60)/24)/365) = 584,942,417.356

IOW if the user allocates a new candidate every single millisecond, non-stop it would take 584 million years to wrap this counter. Previously as a uint32_t it would take 49 days

((((4294967295/1000)/60)/60)/24) = 49.7102696179